modules/renter/siafile/siafile.go

```diff
@@ -0,0 +1,180 @@
+		// and can therefore be easily extended.
+		metadata Metadata
+
+		// pubKeyTable stores the public keys of the hosts this file's pieces are uploaded to.
+		// Since multiple pieces from different chunks might be uploaded to the same host, this
+		// allows us to deduplicate the rather large public keys.
+		pubKeyTable []types.SiaPublicKey
+
+		// chunks are the chunks the file was split into.
+		chunks []Chunk
+
+		// utility fields. These are not persisted.
+		deleted     bool
+		erasureCode modules.ErasureCoder
+		mu          sync.RWMutex
+		uid         string
+	}
+
+	// Chunk represents a single chunk of a file on disk
+	Chunk struct {
+		// erasure code settings.
+		//
+		// erasureCodeType specifies the algorithm used for erasure coding
+		// chunks. Available types are:
+		//   0 - Invalid / Missing Code
+		//   1 - Reed Solomon Code
+		//
+		// erasureCodeParams specifies possible parameters for a certain
+		// erasureCodeType. Currently params will be parsed as follows:
+		//   Reed Solomon Code - 4 bytes dataPieces / 4 bytes parityPieces
+		//
+		erasureCodeType   [4]byte
+		erasureCodeParams [8]byte
+
+		// extensionInfo is some reserved space for each chunk that allows us
+		// to indicate if a chunk is special.
+		extensionInfo [16]byte
+
+		// pieces are the pieces of the file the chunk consists of.
+		pieces [][]Piece
+	}
+
+	// Piece represents a single piece of a chunk on disk
+	Piece struct {
+		KeyNonce   [4]byte            // nonce used for encrypting the piece
+		HostPubKey types.SiaPublicKey // public key of the host
+		MerkleRoot crypto.Hash        // merkle root of the piece
+	}
+)
+
+// New create a new SiaFile.
+func New(siaPath string, erasureCode modules.ErasureCoder, pieceSize, fileSize uint64) *SiaFile {
+	file := &SiaFile{
+		metadata: Metadata{
+			fileSize:  int64(fileSize),
+			masterKey: crypto.GenerateTwofishKey(),
+			pieceSize: pieceSize,
+			siaPath:   siaPath,
+		},
+		erasureCode: erasureCode,
+		uid:         base32.StdEncoding.EncodeToString(fastrand.Bytes(20))[:20],
+	}
+	chunks := make([]Chunk, file.NumChunks())
+	for i := range chunks {
+		chunks[i].erasureCodeType = [4]byte{0, 0, 0, 1}
+		binary.LittleEndian.PutUint32(chunks[i].erasureCodeParams[0:4], uint32(erasureCode.MinPieces()))
+		binary.LittleEndian.PutUint32(chunks[i].erasureCodeParams[4:8], uint32(erasureCode.NumPieces()-erasureCode.MinPieces()))
+		chunks[i].pieces = make([][]Piece, erasureCode.NumPieces())
+	}
+	file.chunks = chunks
+	return file
+}
+
+// AddPiece adds an uploaded piece to the file. It also updates the host table
+// if the public key of the host is not aleady known.
+func (sf *SiaFile) AddPiece(pk types.SiaPublicKey, chunkIndex, pieceIndex uint64, merkleRoot crypto.Hash) error {
+	sf.mu.Lock()
+	defer sf.mu.Unlock()
+
+	// Get the index of the host in the public key table.
+	tableIndex := -1
+	for i, hpk := range sf.pubKeyTable {
+		if reflect.DeepEqual(hpk, pk) {
```

Comment:
Reflection is Overkill and always makes me uneasy; I'd prefer `hpk.Algorithm == pk.Algorithm && bytes.Equal(hpk.Key, pk.Key)`.